Request to evaluate added utils module

Anupama-Github opened the pull request:
@bravegnu Could you evaluate the code for utils module and let me know the corrections?
Thank you

bravegnu commented:
First off, please do not include any generated file in your pull request. Please remove the `_build` folder, backup files ending in "~", etc.

shrayasr commented:
Another key thing to keep in mind is writing good commit messages. Here are some links w.r.t that: 
- https://gist.github.com/matthewhudson/1475276
- http://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html

Also would recommend not to have the `dictionary.txt` added since the file is **huge**

bravegnu commented on README.rst:
```diff
@@ -28,15 +28,8 @@ The package can be installed using the following command ::
-   $ source env/bin/activate
-
-You can now install the package in development mode using the following command ::
+After cloning the repository, install the package in development mode
```
This I guess is unintentional.

bravegnu commented on doc/changelog.rst:
```diff
@@ -1,6 +1,7 @@
 Changelog
 =========
 
+* :feature: '2' Added utils module
```
Please note the spacing and quoting.

```
:feature:`2` Added utils module
```

bravegnu commented on libcipher/utils.py:
```diff
@@ -0,0 +1,46 @@
+    def __init__(self):
+        self.UPPERLETTERS = 'ABCDEFGHIJKLMNOPQRSTUVWXYZ'
+        lower = self.UPPERLETTERS.lower()
+        self.LETTERS_AND_SPACE = self.UPPERLETTERS + lower + ' \t\n'
```
This can be obtained from string module.

```
string.ascii_letters + string.whitespace
```

bravegnu commented on libcipher/utils.py:
```diff
@@ -0,0 +1,46 @@
+        self.UPPERLETTERS = 'ABCDEFGHIJKLMNOPQRSTUVWXYZ'
+        lower = self.UPPERLETTERS.lower()
+        self.LETTERS_AND_SPACE = self.UPPERLETTERS + lower + ' \t\n'
+        f = open('dictionary.txt')
```
Use `with` statement. Also need to handle `IOError`.

bravegnu commented on libcipher/utils.py:
```diff
@@ -0,0 +1,46 @@
+        lower = self.UPPERLETTERS.lower()
+        self.LETTERS_AND_SPACE = self.UPPERLETTERS + lower + ' \t\n'
+        f = open('dictionary.txt')
+        self.englishWords = {}
```
How about using a `set` instead.

bravegnu commented on libcipher/utils.py:
```diff
@@ -0,0 +1,46 @@
+        f.close()
+        self.ENGLISH_WORDS = self.englishWords
+
+    def getEnglishCount(self, message):
```
Use `_` to separate words, instead of camel case for funtions, methods and local variables. PEP8 recommendation.

bravegnu commented on libcipher/utils.py:
```diff
@@ -0,0 +1,46 @@
+        return ''.join(lettersOnly)
+
+    def is_english(self, message, wordPercentage=25, letterPercentage=85):
+        """Takes input, returns True for English input, else returns False"""
```
Use Google Docstring convention for public methods.

bravegnu commented on tests/englishcheck_test.py:
```diff
@@ -0,0 +1,20 @@
+"""Unit tests for english language checker"""
```
Name this `utils_test.py`

Anupama-Github commented:
@shrayasr : Thank you for the links. Will try and improve my commit messages
Regarding inappropriateness of dictionary.txt, would it be better if I used a package such as langdetect(https://pypi.python.org/pypi/langdetect/1.0.1)

Thank you

Anupama-Github commented:
@bravegnu : Thank you for showing the areas needing correction. I have attempted to fix the issues you pointed out. Could you let me know if this would be fine or as suggested by @shrayasr , do I look for an alternative to including dictionary.txt file

Anupama-Github commented:
@bravegnu : Thank you for showing the areas needing correction. I have attempted to fix the issues you pointed out. Could you let me know if this would be fine or as suggested by @shrayasr , do I look for an alternative to including dictionary.txt file

bravegnu commented:
Thanks for updating the code. But this time, you seem to have added back the editor backup files. Here are some guidelines.
1. Avoid a blind `git commit -a`, instead add files explicitly, there are only a few files in this case.
2. Even if you did a `git commit -a`, you need to review the files going into the commit. In the case of git, never provide the commit log in the command line, using the `-m` option. That way you can get to provide a longer commit message, and git shows you all the files going into the commit, in the editor window, so that you can review it.
3. And finally make sure you review your changes before pushing them for review.

That said, it is good to see that you are actively working on improving your code. :smile:

bravegnu commented on libcipher/utils.py:
```diff
@@ -0,0 +1,63 @@
+    def __init__(self):
+       
+        self.LETTERS_AND_SPACE = string.ascii_letters + string.whitespace
+        #f = open('/home/travis/build/Anupama-Github/libcipher/libcipher/dictionary.txt')
```
If it is not required, delete it.

bravegnu commented on libcipher/utils.py:
```diff
@@ -0,0 +1,63 @@
+       
+        self.LETTERS_AND_SPACE = string.ascii_letters + string.whitespace
+        #f = open('/home/travis/build/Anupama-Github/libcipher/libcipher/dictionary.txt')
+        f = open('/home/travis/build/Chennaipy/libcipher/libcipher/dictionary.txt')
```
Hardcoding paths in the library is _really_ crazy. Pass the file object as argument to `__init__()` instead.

shrayasr replied:
Replace the argument `filename` with `dict_file`. 

The term Vijay used there is jargon for [`sed`](http://en.wikipedia.org/wiki/Sed)'s syntax for substitution which is `s/<old word>/<new word/`

bravegnu commented on libcipher/utils.py:
```diff
@@ -0,0 +1,63 @@
+        #f = open('/home/travis/build/Anupama-Github/libcipher/libcipher/dictionary.txt')
+        f = open('/home/travis/build/Chennaipy/libcipher/libcipher/dictionary.txt')
+        self.english_words = {}
+        for word in f.read().split('\n'):
```
`for word in f:` will do. Though `f` would be a bad choice for a variable name.

bravegnu commented on libcipher/utils.py:
```diff
@@ -0,0 +1,63 @@
+        """Checks to see if given text is in English and 
+           returns the result.
+        Args:
+        self (object): Default argument in methods of a class.
```
Don't document self.

bravegnu replied:
Lines below `Args:` have to be indented by two  spaces.

bravegnu commented on libcipher/utils.py:
```diff
@@ -0,0 +1,63 @@
+        percentage of English alphabets in a word. The 
+        default value is 85 
+        Returns:
+        bool: True if message is in English, False otherwise.
```
Lines below `Returns:` have to be indented by two spaces.

bravegnu commented on setup.py:
```diff
@@ -12,7 +12,7 @@ def readme():
       author='Vijay Kumar B.',
       author_email='vijaykumar@bravegnu.org',
-      url='http://github.com/chennaipy/libcipher',
+      url='http://github.com/Anupama-Github/libcipher',
```
Bad idea. :smile:

Anupama-Github replied:
Thank you for pointing this out. It was done to see if it related to my previous builds failure and I forgot to undo it. Will do so

Anupama-Github commented:
@bravegnu : I have made the changes . Could you review ?
Thank you

shrayasr commented on tests/utils_test.py:
```diff
@@ -0,0 +1,22 @@
+import unittest
+import libcipher.utils as utils
+
+filepath = '/home/travis/build/Chennaipy/libcipher/libcipher/dictionary.txt'
```
Hardcoding paths here is a very bad idea since i'll not be able to run this on my machine since this path doesn't exist for me, right? 

Tests aren't only for Travis, they are to be run by every developer.

Anupama-Github replied:
Would this be a better way to handle the filepath location variable:
filepath = os.path.dirname(os.path.abspath(inspect.getfile(utils)))
Could you tell me the correct way of doing so?
Thank you

bravegnu replied:
@Anupama-Github how about changing the API to accept a file object instead of a filename? This will make the API more flexible, and in the test case we could pass in a StringIO object, with the dictionary contents.

shrayasr replied:
:+1: IIRC, I think even [NLTK](http://www.nltk.org/) does the same kind of thing, the dictionaries are separate from the actual toolkit itself and have to be downloaded separately before NLTK can use them.

Anupama-Github replied:
@bravegnu  I am not sure if I understood what you said. I have locally modified utils_test.py to get the filepath and create file objects using

```
filepath = os.path.dirname(os.path.abspath(inspect.getfile(utils)))
fileobject = open(filepath+'/dictionary.txt')
```

I pass fileobject to utils.py
The tests have run fine. Is this what you meant by "changing the API to accept a file object instead of a filename" ?
Thank you

shrayasr replied:
You can find more about File objects in the [python 2 documentation](https://docs.python.org/2/library/stdtypes.html#bltin-file-objects) and in the [python 3 documentation](https://docs.python.org/3/glossary.html#term-file-object)

The idea is to keep the act of opening of the file and storing its content into a `File` object outside of the library to make it independent of where the `dictionary.txt` file is stored.

bravegnu replied:
@Anupama-Github I did intend that you pass the file object, returned from `open()` to the `EnglishChecker` constructor. But in the test case, you can avoid having a dictionary file, by using an in memory file object, as shown below.

``` python
    def get_marys_dictionary():
        english_words = ["had", "a", "little", "lamb"]
        dict_file = io.StringIO("\n".join(english_words))
        return dict_file
    ...

    class EnglishCheckerTestCase(unittest.TestCase):
        def test_for_english(self):
            dict_file = get_marys_dictionary()
            checker = utils.EnglishChecker(dict_file)
```

For more details about StringIO see https://docs.python.org/3/library/io.html

Anupama-Github replied:
@bravegnu , @shrayasr: I still am unable to follow. If I need to put the contents of dictionary file into the `english_words` list, it would have 45333 words. Should I do that, and then use the StringIO class to create the object dict_file and pass this to the 
`Englishchecker` constructor?
Thank you

bravegnu replied:
You just need to have the words "had", "a", "little", "lamb", in the dictionary as indicated above.

Anupama-Github replied:
If that is so, how would a sentence not containing these 4 words be recognized as English?

bravegnu replied:
We are only testing one sentence in that test case. And the dictionary for that sentence is sufficient.

shrayasr commented on libcipher/utils.py:
```diff
@@ -0,0 +1,63 @@
+        
+
+    def get_english_count(self, message):
+        #print (self.english_words)
```
Before you push code, make sure you remove such statements which you'd have used to debug :)

shrayasr commented:
More general points: 
- Also please revert `README.rst`. It looks like you've made some changes unknowingly. 
- `dictionary.txt` **should** go. Although I should admit that I don't know what to substitute this with so i'll leave that decision to @bravegnu 

Great work! :+1:

Anupama-Github commented:
@bravegnu I have used your code to remove dictionary file from utils_test.py . As I still do not understand how different english sentences can be tested, I have used dictionary.txt in the utils file. could you tell me what to use if you want it removed altogether?
Thank you

bravegnu commented:
The `StringIO` is grossly misused in the code. There are two separate problems that we are discussing here.
1. Storing a dictionary file within the code base.
2. Dictionary file for the unit test cases.

For the first problem, the solution is not to store any dictionary at all in the code base. Let the user get his own dictionary, open it and pass the file object in. StringIO, StringIO checks are not required whatsoever.

For the second problem, the StringIO can be used, to create an in memory dictionary file, containing the bare minimum set of words required for the tests to pass / fail.

Actions to be taken:
1. Drop the dictionary file from the code base.
2. Drop the code that reads the above dictionary file. And no default dictionary argument.
3. Drop the checks for StringIO in the `__init__()` method. `StringIO` when passed from the test case appears just like a file object to the code. That is the whole point of using `StringIO`.
4. Drop the code that reads the dictionary into a StringIO in the `__init__()` method.

Anupama-Github commented:
@bravegnu I have done as you said. Is it fine now?
Thank you

shrayasr commented:
@Anupama-Github: The `dictionary.txt` file is still in the list of things to be committed. @bravegnu has also mentioned to _Drop the dictionary file from the code base_.

In addition, Please revert all changes you have made to the `README.rst` file as well.

Anupama-Github commented:
It was careless of me to leave it. I have removed it . Is it fine now?
Thank you

Anupama-Github committed:
Replaced with original README

Anupama-Github committed:
Merge branch 'master' of https://github.com/Anupama-Github/libcipher

Removed dictionary file

bravegnu commented on tests/utils_test.py:
```diff
@@ -0,0 +1,40 @@
+        return dict_file
+
+
+class english_check_testcase(unittest.TestCase):
```
PEP8 recommendation is `EnglishCheckTestCase`.

bravegnu commented on libcipher/utils.py:
```diff
@@ -0,0 +1,61 @@
+           the dictionary file, an exhaustive list of english words
+    """
+
+    def __init__(self, filename):
```
s/filename/dict_file/

Anupama-Github replied:
@bravegnu I am not able to understand this note. Could you explain?
Thank you

bravegnu commented on tests/utils_test.py:
```diff
@@ -0,0 +1,40 @@
+        if sys.version < '3':
+            dict_file = io.StringIO(unicode('\n').join(english_words))
+        else:
+            dict_file = io.StringIO("\n".join(english_words))
```
If this library is about checking for English words, why would we use unicode strings in Python 2?

Anupama-Github replied:
@bravegnu Removing the unicode function caused the build to fail for python 3.2 I do not know why this happens. Can i try this any other way?
Thank you

bravegnu commented on libcipher/utils.py:
```diff
@@ -0,0 +1,61 @@
+        self.filename = filename
+        self.LETTERS_AND_SPACE = string.ascii_letters + string.whitespace
+        content = filename.getvalue()
+        self.ENGLISH_WORDS = content.upper()
```
Two issues here:
1. `.getvalue()` is a StringIO method. You should be using `read()` or better yet `readlines()`.
2. This keeps the whole dictionary as one big string. And later on we are searching through the string. This works, but is grossly inefficient. The dictionay will have one word per line. Read all the lines, convert them to upper case, and then make a it a `set`.

bravegnu commented on libcipher/utils.py:
```diff
@@ -0,0 +1,61 @@
+        num_letters = len(self.remove_non_letters(message))
+        message_letters_percentage = float(num_letters) / len(message) * 100
+        letters_match = message_letters_percentage >= letter_percentage
+        self.filename.close()
```
Close in `__init__()` after reading the data.

Anupama-Github replied:
@bravegnu I have closed it in the test functions. Is that alright? Also the tests run even if I do not close and it says [here](http://stackoverflow.com/questions/11892623/python-stringio-and-compatibility-with-with-statement-context-manager) that one need not close a stringio object. Could you explain?
Thank you

bravegnu commented:
Looks much better now, please have a look at the comments.

Anupama-Github committed:
Changed the way dictionary is read and modified docstring

Anupama-Github committed:
checking to see if code is python 3.2 compatible

Anupama-Github committed:
using if-else based on version for compatibility with python 3.2

Anupama-Github committed:
testing

Anupama-Github committed:
testing

bravegnu commented on libcipher/utils.py:
```diff
@@ -0,0 +1,62 @@
+    """
+
+    def __init__(self, filename):
+        self.filename = filename
```
Just call the argument `dict_file` or `dictionary_file` instead of `filename`. Since it is _not_ a filename, but a file object. Same with the attribute as well.

shrayasr commented on README.rst:
```diff
@@ -72,3 +72,4 @@ Before sending a pull request, please ensure the following:
 * The changelog has been updated, to indicate the change made along
   with the issue no. on GitHub. The changelog is to be in `releases
   format <http://releases.readthedocs.org/en/latest/index.html>`_.
+
```
There still seems to be some change made here @Anupama-Github. Request you to revert this file so that other changes made to `README.rst` will be brought in on merge

Anupama-Github committed:
Update README.rst

Anupama-Github committed:
changed name of file object

Anupama-Github commented:
@bravegnu @shrayasr I have made the changes. Is it fine now?
Thank you

bravegnu commented on libcipher/utils.py:
```diff
@@ -0,0 +1,62 @@
+        self.LETTERS_AND_SPACE = string.ascii_letters + string.whitespace
+        content = ''.join(dict_file.readlines())
+        content = set(content.upper().split())   
+        self.ENGLISH_WORDS = content
```
Not sure, why I would want to read them as lines and join the back! 3 lines above can be replaced with:

```
lines = dict_file.readlines()
self.english_words = set(word.upper().strip() for word in lines)
```

bravegnu commented on tests/utils_test.py:
```diff
@@ -0,0 +1,39 @@
+            dict_file = io.StringIO(unicode('\n').join(english_words))
+        else:
+            dict_file = io.StringIO("\n".join(english_words))
+        return dict_file
```
AFAIK, this is not required. Please recheck.

Anupama-Github replied:
Removing the if condition causes this error in Python 2;
 dict_file = io.StringIO("\n".join(english_words))
TypeError: initial_value must be unicode or None, not str

bravegnu replied:
OK, now I understand the problem, it has got to do with `io.StringIO` class. Will let you know how this can be fixed.

Anupama-Github replied:
Thank you

Anupama-Github commented:
@bravegnu I had updated the code in utils.py to read from dictionary file. I got a Type error when I removed the version check condition in utils_test.py

Anupama-Github committed:
Updated code to read lines from dictionary
